eipw-lint/src/lints/markdown/link_eip.rs

```diff
@@ -0,0 +1,141 @@
+            captures.get(1).map(|m| m.as_str())
+        } else { None };
+
+        if let Some(url_eip_number) = url_eip_number {
+            let section_pattern = r"eip-([^.]*)\.md#.+$";
+            let url_re = Regex::new(section_pattern).map_err(Error::custom)?;
+            let dynamic_pattern = if url_re.is_match(&self.current_link.url) {
+                format!(r"^(EIP|ERC)-{}\s*\S+", regex::escape(&url_eip_number))
+            } else {
+                format!(r"^(EIP|ERC)-{}$", regex::escape(&url_eip_number))
+            };
+            let text_re = Regex::new(&dynamic_pattern).map_err(Error::custom)?;
+        
+            if text_re.is_match(&self.current_link.text) {
+                return Ok(Next::TraverseChildren);
+            };
+        
+            let expected = if url_re.is_match(&self.current_link.url) {
+                format!("[EIP|ERC-{}<section-description>]", url_eip_number)
+            } else {
+                format!("[EIP|ERC-{}]", url_eip_number)
+            };
+        
+            let footer_label = format!("link text should match `{}`", expected);
+        
+            let source = self.ctx.source_for_text(ast.sourcepos.start.line, &self.current_link.text);
+            self.ctx.report(Snippet {
+                title: Some(Annotation {
+                    annotation_type: self.ctx.annotation_type(),
+                    id: Some(self.slug),
+                    label: Some("link text does not match link destination"),
+                }),
+                slices: vec![Slice {
+                    fold: false,
+                    line_start: ast.sourcepos.start.line,
+                    origin: self.ctx.origin(),
+                    source: &source,
+                    annotations: vec![],
+                }],
+                footer: vec![Annotation {
+                    id: None,
+                    annotation_type: AnnotationType::Info,
+                    label: Some(&footer_label),
+                }],
+                opt: Default::default(),
+            })?;
+        }
+
+        Ok(Next::TraverseChildren)
+    }
+}
+
+impl<'a, 'b, 'c> tree::Visitor for Visitor<'a, 'b, 'c> {
+    type Error = Error;
+
+    fn enter_link(&mut self, _: &Ast, link: &NodeLink,) -> Result<Next, Self::Error> {
+        if self.re.is_match(&link.url) {        
+            self.current_link = Link { url: link.url.to_owned(), text: String::new() }; 
+            self.link_depth += 1;  
+        }                          
+        Ok(Next::TraverseChildren)
+    }
+
+    fn depart_link(&mut self, _: &Ast, _: &NodeLink) -> Result<(), Self::Error> {
+        if self.link_depth > 0 {
+            self.link_depth = self.link_depth.checked_sub(1).unwrap();
+        }
+        Ok(())
+    }
+
+    fn enter_text(&mut self, ast: &Ast, txt: &str) -> Result<Next, Self::Error> {
+        if self.link_depth > 0 {
+            self.current_link.text = txt.to_owned();
+            self.check(ast)?;         
+        }
```

Comment:
How does this behave for content like:

```markdown
[**EIP-1**5678](./eip-1.md#rationale)
```

Reply:
It fails. I will add it to the tests. Should it fail?

Reply:
Hm, perhaps I spoke too soon then :rofl:

I was worried that because every text node in the link sets `self.current_link.text`, you could end up in situations where, for example, bold would create two text nodes, breaking the lint. Maybe something like `[**EIP-1**EIP-1](./eip-1.md)`.

If my concerns are unfounded, ignore me!